### PR TITLE
Refactoring endpoints and GameState class

### LINKED ISSUE #72: Change python game_state structure to a class
Right now `game_state` is a dict with lots of helper functions.  It's a prime candidate to become a class, and those helper functions will become methods.

---

Implementation:

create a `class GameState(object)`.  It should have the following methods:
- `@classmethod` `load(slug)`
- `save()`
- `get_seed_count()`
- `get_resource_count()`
- `get_plot(x, y)`
- `has_planted_crops()`
- `initialize()`
- `get_unlocked_plot_count()`
- `check_password(password)`
- `jsonify()`

The dict storing the data should be private to the class (`self._data = {...}`)

Existing code should be refactored to no longer need direct dict access.

### LINKED ISSUE #73: Refactor endpoints to separate game logic
An endpoint function (one annotated with `@app.route`) should only do the following:

- extract data from the request object (pulling from `request.json` and separating out the useful pieces)
- identify what's being requested by the client
- call a function or few functions to do what's requested
- combine the results of the functions into a response the client will understand

Clean up the existing endpoints so they're easier to read and offload data manipulation to helper functions.

### PR DESCRIPTION
**Pull request by bjones3**
Fixes #72
Fixes #73

### CODE REVIEW

File: webserver.py [hunk 1]
```diff
@@ -176,216 +308,76 @@ def game_config():
...
 def recipe():
-    data = request.json  # data={slug}
-    game_state = load_state(data['slug'])
+    data = request.json  # data={slug,password}
+    game_state = GameState.load(data['slug'], data['password'])
 
     known_recipes = {}
     for recipe_id in RECIPE_CONFIG['recipes']:
-        if recipe_id in game_state['recipes']:
+        if recipe_id in game_state.data['recipes']:
```
Comment (phblj) [inline webserver.py]: Instead of accessing `data`, you can hide the implementation inside your class and create an accessor method.

The big reason to do this is to make altering your class behavior later easier.  If you have a method `get_recipes` or `is_known_recipe(recipe_id)` it's clear what it's supposed to do.  Giving general access to the full `data` field outside the class means you have to support whatever way you want to use the class.  In this case, the handler code has to know how game_states are structured, and game_states have to expose this structure so others can manipulate it.

I'd rather have this line read,

`if game_state.is_known_recipe(recipe_id):`

or if that method's not general-purpose enough for you,

`if recipe_id in game_state.get_recipes():`

File: webserver.py [hunk 2]
```diff
@@ -127,7 +245,7 @@ def get_admin_data():
 
 def make_response(game_state, message=None, recipes=None):
     response = {
-        'state': game_state
+        'state': game_state.data
```
Comment (phblj) [inline webserver.py]: rather than directly access the data, put it behind a method, "get_client_data"

Down the road, this will let us specify exactly which parts of the game state get sent to the client.  For instance, there's not a good reason to send the password back, ever.  So if you want to send all the data except the `password` field, this would be a good place.

**Review by phblj — CHANGES_REQUESTED** [review]

File: webserver.py [hunk 3]
```diff
@@ -144,6 +262,20 @@ def generate(seed_id):
...
+    while generate(seed_id) == 1:
+        bonus += 1
+    return bonus
+
+
+def valid_chars(slug):
+    regex = re.compile('^[a-z]+$')
+    x = regex.match(slug)
+    if x is False:
+        raise Exception("Invalid characters")
```
Comment (phblj) [inline webserver.py]: We can do better than generic Exceptions now.  This seems like a BadRequest to me, but take a look at
http://werkzeug.pocoo.org/docs/0.11/exceptions/

And use an appropriate error.

File: webserver.py [hunk 4]
```diff
@@ -41,61 +42,178 @@ def get_conn():
...
+                cursor.execute('INSERT INTO game_states (slug, game_state) VALUES (%s, %s) '
+                               'ON CONFLICT (slug) DO UPDATE SET game_state=%s;',
+                               [slug, json_game_state, json_game_state])
+
+    # add missing seed counts, resources, and plots to game_state
+    # unlock first four plots
+    # check to never run out of money
+    def initialize(self):
+        for seed_id in GAME_CONFIG['seeds']:
+            if self.data['seedCounts'].get(seed_id) is None:
```
Comment (phblj) [inline webserver.py]: Now that we've got a class, this can be cleaned up a lot to make it more readable.

Rather than `if self.data['seedCounts'].get(seed_id) is None...` checks, add a `get_seed_count(seed_id)` method that returns 0 if the seed's not found.  Then this code becomes much cleaner.

File: webserver.py [hunk 5]
```diff
@@ -41,61 +42,178 @@ def get_conn():
...
+                self.data['seedCounts'][seed_id] = 0
+        for resource in GAME_CONFIG['starting_resources']:
+            if self.data['resources'].get(resource) is None:
+                self.data['resources'][resource] = GAME_CONFIG['starting_resources'][resource]
+        for i in range(GAME_CONFIG['field_width']):
```
Comment (phblj) [inline webserver.py:100]: The `self.data.get('plot'...` can be simplified a little, but leave it until you tackle #77.  That's going to make things a lot easier to read

File: webserver.py [hunk 6]
```diff
@@ -41,61 +42,178 @@ def get_conn():
...
+        total_seed_count = 0
+        total_crops_planted = 0
+        for seed_id in GAME_CONFIG['seeds']:
+            total_seed_count += self.data['seedCounts'][seed_id]
+        for i in range(GAME_CONFIG['field_width']):
+            for j in range(GAME_CONFIG['field_height']):
+                if self.has_planted_crop(i, j):
+                    total_crops_planted += 1
+        if total_seed_count == 0 and total_crops_planted == 0:
+            if self.data['resources']['cash'] < RECIPE_CONFIG['recipes']['a']['cashCost']:
```
Comment (phblj) [inline webserver.py]: rather than using `'a'` here, check cash against the cost whatever's set as `firstSeed` in the config.  This way we're not tied to always having a potato if we want to change the planting list

File: webserver.py (hunk 4 above)
Comment (bjones3) [reply webserver.py]: Every way in which I tried simplifying this with another function just ended up making it more confusing.

File: webserver.py [hunk 7]
```diff
@@ -46,6 +46,10 @@ class GameState(object):
     def __init__(self, json_data):
         self.data = json_data
 
+    def get_client_data(self):
+        del self.data['password']
```
Comment (phblj) [inline webserver.py]: careful here.  You're actually removing the 'password' key from the data.  If you were to later save this game_state, you'd be storing a game without a password, so the user could never log back in.

Another problem is that you're exposing the underlying data itself.  It's not hard to imaging down the road having code like this:

```
client_data = game_state.get_client_data()
clients_seeds = client_data['seeds']
msg = ""
while client_seeds:
    msg += "You have %s %s seeds.\n" % client_seeds.pop_item()
print msg
...
client_data.save()
```

The above code would, as a side effect, empty the player's inventory.  The person writing the code probably wouldn't expect that to happen.


A strategy to deal with this is called a "defensive copy":  http://www.javapractices.com/topic/TopicAction.do?Id=15
Basically you make a copy of the data you care about and hand that out.  The caller can perform whatever changes they want on that copy without affecting your original.

python as a module `copy` with a method `deepcopy` that might come in handy here

**Review by phblj — CHANGES_REQUESTED** [review]

File: webserver.py (hunk 4 above)
Comment (phblj) [reply webserver.py]: You're right-- this is getting a little ahead of ourselves (you'll do this in #77) but this is a good place to talk about it.  Don't worry about changing it in this PR, but:

this chunk of code:
```
        for seed_id in GAME_CONFIG['seeds']:
            if self.data['seedCounts'].get(seed_id) is None:
                self.data['seedCounts'][seed_id] = 0
```
only exists to initialize the game_state to 0s when a new seed is introduced.  But we can make a rule that says, "when a seed count doesn't exist, we consider it a 0":

```
def get_seed_count(self, seed_id):
    count = self.data['seedCounts'].get(seed_id)
    if count is None:
        return 0
    return count
```

now all your `self.data['seedCounts'].get(seed_id)` calls can just be `self.get_seed_count(seed_id)`.  More importantly, you don't run the risk that if you ever write code where `initialize` fails to get called at the right time, this part won't break-- you'll never end up with a `None` seed_count

**Comment by phblj** [comment]
Take a look at the "Defensive copy" change on `get_client_data`.

Read (but you don't need to change) my comment on seedCount getting

I'll fix the test-- it's failing because password is missing (by design)
